type_description_interfaces/msg/FieldType.msg

```diff
@@ -0,0 +1,149 @@
+uint8 FIELD_TYPE_INT64_BOUNDED_SEQUENCE = 71
+uint8 FIELD_TYPE_UINT64_BOUNDED_SEQUENCE = 72
+uint8 FIELD_TYPE_FLOAT_BOUNDED_SEQUENCE = 73
+uint8 FIELD_TYPE_DOUBLE_BOUNDED_SEQUENCE = 74
+uint8 FIELD_TYPE_LONG_DOUBLE_BOUNDED_SEQUENCE = 75
+uint8 FIELD_TYPE_CHAR_BOUNDED_SEQUENCE = 76
+uint8 FIELD_TYPE_WCHAR_BOUNDED_SEQUENCE = 77
+uint8 FIELD_TYPE_BOOLEAN_BOUNDED_SEQUENCE = 78
+uint8 FIELD_TYPE_BYTE_BOUNDED_SEQUENCE = 79
+uint8 FIELD_TYPE_STRING_BOUNDED_SEQUENCE = 80
+uint8 FIELD_TYPE_WSTRING_BOUNDED_SEQUENCE = 81
+uint8 FIELD_TYPE_FIXED_STRING_BOUNDED_SEQUENCE = 82
+uint8 FIELD_TYPE_FIXED_WSTRING_BOUNDED_SEQUENCE = 83
+uint8 FIELD_TYPE_BOUNDED_STRING_BOUNDED_SEQUENCE = 84
+uint8 FIELD_TYPE_BOUNDED_WSTRING_BOUNDED_SEQUENCE = 85
+
+# Unbounded Sequence Types
+uint8 FIELD_TYPE_NESTED_TYPE_UNBOUNDED_SEQUENCE = 96
+uint8 FIELD_TYPE_INT8_UNBOUNDED_SEQUENCE = 97
+uint8 FIELD_TYPE_UINT8_UNBOUNDED_SEQUENCE = 98
+uint8 FIELD_TYPE_INT16_UNBOUNDED_SEQUENCE = 99
+uint8 FIELD_TYPE_UINT16_UNBOUNDED_SEQUENCE = 100
+uint8 FIELD_TYPE_INT32_UNBOUNDED_SEQUENCE = 101
+uint8 FIELD_TYPE_UINT32_UNBOUNDED_SEQUENCE = 102
+uint8 FIELD_TYPE_INT64_UNBOUNDED_SEQUENCE = 103
+uint8 FIELD_TYPE_UINT64_UNBOUNDED_SEQUENCE = 104
+uint8 FIELD_TYPE_FLOAT_UNBOUNDED_SEQUENCE = 105
+uint8 FIELD_TYPE_DOUBLE_UNBOUNDED_SEQUENCE = 106
+uint8 FIELD_TYPE_LONG_DOUBLE_UNBOUNDED_SEQUENCE = 107
+uint8 FIELD_TYPE_CHAR_UNBOUNDED_SEQUENCE = 108
+uint8 FIELD_TYPE_WCHAR_UNBOUNDED_SEQUENCE = 109
+uint8 FIELD_TYPE_BOOLEAN_UNBOUNDED_SEQUENCE = 110
+uint8 FIELD_TYPE_BYTE_UNBOUNDED_SEQUENCE = 111
+uint8 FIELD_TYPE_STRING_UNBOUNDED_SEQUENCE = 112
+uint8 FIELD_TYPE_WSTRING_UNBOUNDED_SEQUENCE = 113
+uint8 FIELD_TYPE_FIXED_STRING_UNBOUNDED_SEQUENCE = 114
+uint8 FIELD_TYPE_FIXED_WSTRING_UNBOUNDED_SEQUENCE = 115
+uint8 FIELD_TYPE_BOUNDED_STRING_UNBOUNDED_SEQUENCE = 116
+uint8 FIELD_TYPE_BOUNDED_WSTRING_UNBOUNDED_SEQUENCE = 117
+
+# Identifying number for the type of the field, using one of the above constants.
+uint8 type_id
+
+# Only used when the type is an array or a bounded sequence.
+# In the case of an array, this is the fixed length of the array.
+# In the case of a bounded sequence, this is the maximum size of the sequence.
+# In all other cases this field is unused.
+uint64 length
+
+# Only used when the type is a fixed or bounded string/wstring, or a array/sequence of those.
+# In the case of a fixed string/wstring, it is the fixed length of the string.
+# In the case of a bounded string/wstring, it is the maximum length of the string.
+# In the case of an array/sequence of fixed string/wstring, it is the fixed length of the strings.
+# In the case of an array/sequence of bounded string/wstring, it is the maximum length of the strings.
+# It is not currently possible to have different string lengths per element in the array/sequences.
+uint64 string_length
+
+# Only used when the type is a nested type or array/sequence of nested types.
+string nested_type_name
```

Comment:
should this be bounded string? do type names have a bound in ros2?

Reply:
I think there's an implicit limit because of the max length of a file name as supported by the OS (~255 - 4 = 251 max), but this does not extend to IDL files AFAIK.

Reply:
I'm also not sure about this, but yes I could see this either being a fixed (I know we don't support that atm) or bounded sized string.

Reply:
It is bound to 256 bytes, which must include the null terminator, so 255 characters total. See section 9.6.3.2.1 of the DDSI-RTPS spec, table 9.18: https://www.omg.org/spec/DDSI-RTPS/2.5/PDF

I can go ahead and bound this, but like we talked about off-line, bounding just this field doesn't get us much, if we don't also manage to bound the other strings in these messages.

Reply:
I just re-checked and I could not find any documentation about a limit to the type name length in ROS 2's design documents, so that's something we'll have to follow up on in the future, but is out of scope for this pr. I've added notes about it in commit cd119a6.

I could see us further limiting these type name lengths, e.g. so that we can concatenate the type hash with it in some middlewares. On the other hand, those middlewares might not have a 255 cap themselves, allowing us to concatenate the type hash to a ~255 character type name and still not have a problem. These fields are specific the name without the type hash concatenated.

Also, there are still unbounded strings for the field names and the default value, which would need to be bounded to achieve some real benefits here.